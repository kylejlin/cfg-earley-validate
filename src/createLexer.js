const createLexer = (tokenTypes) => {
  tokenTypes = tokenTypes.map((type) => {
    const { name, rule, value, isOmitted } = type;
    const caretRule = new RegExp('^' + rule.source);
    if (caretRule.test('')) {
      throw new TypeError('Rule ' + rule.source + ' for token type ' + name + ' matched the empty string. Rules cannot match the empty string.');
    }
    return {
      name,
      rule: caretRule,
      value: 'function' === typeof value ? value : null,
      isOmitted,
    };
  });

  return (string) => {
    const tokens = [];
    let line = 1;
    let column = 0;
    while (string.length > 0) {
      let hasFoundRule = false;

      for (const type of tokenTypes) {
        const { name, rule, value: valueMapper, isOmitted } = type;
        const match = string.match(rule);
        if (match === null) {
          continue;
        }
        const [firstMatch] = match;

        string = string.slice(firstMatch.length);

        const startLine = line;
        const startColumn = column;
        const newlines = firstMatch.split('\n').length - 1;
        if (newlines > 0) {
          line += newlines;
          column = firstMatch.split('\n').slice(-1)[0].length;
        } else {
          column += firstMatch.length;
        }

        if (!isOmitted) {
          const value = valueMapper === null
            ? firstMatch
            : valueMapper(firstMatch);
          tokens.push({
            name: name,
            value,
            start: [startLine, startColumn],
            end: [line, column],
          });
        }

        hasFoundRule = true;
        break;
      }

      if (!hasFoundRule) {
        throw new SyntaxError('Cannot tokenize: ' + string);
      }
    }
    return tokens;
  };
};

export default createLexer;
