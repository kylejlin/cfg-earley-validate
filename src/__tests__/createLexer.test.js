import createLexer from '../createLexer';

test('it lexes correctly', () => {
  const tokenTypes = [
    {
      name: 'LPAREN',
      rule: /\(/,
    },
    {
      name: 'RPAREN',
      rule: /\)/,
    },
    {
      name: 'NUMBER',
      rule: /\d+(?:\.\d+)?/,
      value: parseFloat,
    },
    {
      name: 'STRING',
      rule: /"(?:[^"\\]|\\.)*"/,
      value: (value) => value.slice(1, -1),
    },
    {
      name: 'IDENTIFIER',
      rule: /[a-zA-Z_][a-zA-Z_0-9]*/,
    },
    {
      isOmitted: true,
      name: 'WHITESPACE',
      rule: /[ \t\n]+/,
    }
  ];
  const lex = createLexer(tokenTypes);
  const code = '(foo 42 (baz "bar"))';
  expect(lex(code)).toMatchSnapshot();
});
