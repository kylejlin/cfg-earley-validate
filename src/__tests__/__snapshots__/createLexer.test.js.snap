// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it lexes correctly 1`] = `
Array [
  Object {
    "end": Array [
      1,
      1,
    ],
    "name": "LPAREN",
    "start": Array [
      1,
      0,
    ],
    "value": "(",
  },
  Object {
    "end": Array [
      1,
      4,
    ],
    "name": "IDENTIFIER",
    "start": Array [
      1,
      1,
    ],
    "value": "foo",
  },
  Object {
    "end": Array [
      1,
      7,
    ],
    "name": "NUMBER",
    "start": Array [
      1,
      5,
    ],
    "value": 42,
  },
  Object {
    "end": Array [
      1,
      9,
    ],
    "name": "LPAREN",
    "start": Array [
      1,
      8,
    ],
    "value": "(",
  },
  Object {
    "end": Array [
      1,
      12,
    ],
    "name": "IDENTIFIER",
    "start": Array [
      1,
      9,
    ],
    "value": "baz",
  },
  Object {
    "end": Array [
      1,
      18,
    ],
    "name": "STRING",
    "start": Array [
      1,
      13,
    ],
    "value": "bar",
  },
  Object {
    "end": Array [
      1,
      19,
    ],
    "name": "RPAREN",
    "start": Array [
      1,
      18,
    ],
    "value": ")",
  },
  Object {
    "end": Array [
      1,
      20,
    ],
    "name": "RPAREN",
    "start": Array [
      1,
      19,
    ],
    "value": ")",
  },
]
`;
